documentation/docs/14-content-security-policy.md

```diff
@@ -0,0 +1,49 @@
+---
+title: Content Security Policy
+---
+
+At the moment, SvelteKit supports adding Content Security Policy via hooks. In environments with a runtime, HTTP headers can be added to the response object.
+
+However, SvelteKit also requires some small pieces of inline JavaScript in order for hydration to work. To avoid using `'unsafe-inline'` (which, as the name suggests, should be avoided), SvelteKit can be configured to inject CSP nonces into the HTML it generates.
+
+The nonce value is availiable to hooks as `request.locals.nonce`. A basic CSP handler hook might then look like this:
+
+```javascript
+export async function handle ({ request, resolve }) {
+  const directives = {
+    'default-src': ["'self'", 'static.someotherdomain.com'],
+    'script-src': ["'strict-dynamic'"],
+    'style-src': ["'self'"]
+  };
+  const response = await resolve(request);
+
+  const nonce = request.locals.nonce;
+
+  directives['script-src'].push(`'nonce-${nonce}'`);
+  directives['style-src'].push(`'nonce-${nonce}'`);
+
+  if (process.env.NODE_ENV === 'development') {
+    directives['style-src'].push('unsafe-inline')
+  }
+
+  const csp = Object.entries(directives)
+    .map(([key, arr]) => key + ' ' + arr.join(' '))
+    .join('; ');
+
+  return {
+    ...response,
+    headers: {
+      ...response.headers,
+      'Content-Security-Policy': csp
```

Comment:
Doing this in the `handle` hook would end up applying to endpoints as well, which probably isn't quite what we want. We should probably do this in the layout instead. Of course that would require https://github.com/sveltejs/kit/issues/909 to be implemented first. That issue is tagged to the 1.0 milestone, so hopefully we'll get to it before too too long

Reply:
While this is true (and will be relevant later) we can't document for a feature that doesn't exist yet 😄 

Setting a CSP on an endpoint response probably won't _hurt_ (it'll most likely just be discarded), but you can also check if the response type is "text/html" first. It invokes the handler at every endpoint call, but won't add headers to them. I should maybe add that to the example.

I've been thinking a little about a more permanent home for CSP headers in a more built-in way, but that's a different discussion.

Reply:
Actually, doing this in a layout requires more than just the ability to set headers. It also requires access to `request.locals` in order to put the nonce into the header, which is problematic I think.

Reply:
Ok, yeah, let's do it in `handle`. But can we check if the `content-type` is `text/html` and only do it then?

Reply:
Yeah that'll work. In fact I already added that to the example :)